File: system/diagnostic_graph_aggregator/script/dump.py
#!/usr/bin/env python3

# Copyright 2024 The Autoware Contributors
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import argparse

from diagnostic_msgs.msg import DiagnosticStatus
import rclpy
import rclpy.node
from tier4_system_msgs.msg import DiagnosticGraph


def print_table(lines: list, header: list):
    widths = [0 for _ in range(len(header))]
    lines.insert(0, header)
    for line in lines:
        widths = map(max, widths, map(len, line))
    widths = list(widths)
    lines.insert(0, ["-" * w for w in widths])
    lines.insert(2, ["-" * w for w in widths])
    for line in lines:
        line = map(lambda v, w: f"{v:{w}}", line, widths)
        line = " | ".join(line)
        print(f"| {line} |")


def get_level_text(level: int):
    if level == DiagnosticStatus.OK:
        return "OK"
    if level == DiagnosticStatus.WARN:
        return "WARN"
    if level == DiagnosticStatus.ERROR:
        return "ERROR"
    if level == DiagnosticStatus.STALE:
        return "STALE"
    return "-----"


class NodeData:
    def __init__(self, index, node):
        self.index = index
        self._node = node

    @property
    def level(self):
        return get_level_text(self._node.status.level)

    @property
    def name(self):
        return self._node.status.name

    @property
    def links(self):
        return " ".join(str(link.index) for link in self._node.links)

    @property
    def line(self):
        return [str(self.index), self.level, self.name, self.links]


class DumpNode(rclpy.node.Node):
    def __init__(self, args):
        super().__init__("dump_diagnostic_graph")
        self.sub = self.create_subscription(DiagnosticGraph, args.topic, self.callback, 1)

    def callback(self, msg):
        nodes = [NodeData(index, node) for index, node in enumerate(msg.nodes)]
        table = [node.line for node in nodes]
        print_table(table, ["index", "level", "name", "links"])


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--topic", default="/diagnostics_graph")
    args, unparsed = parser.parse_known_args()

    try:
        rclpy.init(args=unparsed)
        rclpy.spin(DumpNode(args))
        rclpy.shutdown()
    except KeyboardInterrupt:
        pass

File: system/diagnostic_graph_utils/src/node/dump.cpp
// Copyright 2024 The Autoware Contributors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "dump.hpp"

#include <algorithm>
#include <iomanip>
#include <iostream>
#include <memory>
#include <sstream>
#include <vector>

namespace diagnostic_graph_utils
{

DumpNode::DumpNode(const rclcpp::NodeOptions & options) : Node("dump", options)
{
  using std::placeholders::_1;
  sub_graph_.register_create_callback(std::bind(&DumpNode::on_create, this, _1));
  sub_graph_.register_update_callback(std::bind(&DumpNode::on_update, this, _1));
  sub_graph_.subscribe(*this, 1);
}

void DumpNode::on_create(DiagGraph::ConstSharedPtr graph)
{
  // Initialize table line information.
  const std::string title_index = "No";
  const std::string title_level = "Level";
  const std::string title_path = "Path";
  const std::string title_type = "Type";
  const std::string title_link = "Children";

  // Cache units.
  const auto units = graph->units();

  // Assign merged units index.
  int table_index = 0;
  for (const auto & unit : units) {
    TableLine line;
    line.index = ++table_index;
    table_.emplace(unit, line);
  }

  // Create link index text. Use text2 as a temporary variable.
  for (const auto & unit : units) {
    std::string links;
    for (const auto & child : unit->children()) {
      links += std::to_string(table_.at(child.unit).index) + " ";
    }
    if (!links.empty()) {
      links.pop_back();
    }
    table_.at(unit).text2 = links;
  }

  // Calculate table cell width.
  const auto width_index = std::max(title_index.length(), std::to_string(units.size()).length());
  const auto width_level = title_level.length();
  auto width_path = title_path.length();
  auto width_type = title_type.length();
  auto width_link = title_link.length();
  for (const auto & unit : units) {
    const auto & line = table_.at(unit);
    width_path = std::max(width_path, unit->path().length());
    width_type = std::max(width_type, unit->type().length());
    width_link = std::max(width_link, line.text2.length());
  }

  // Create table lines.
  for (const auto & unit : units) {
    auto & line = table_.at(unit);
    std::ostringstream text1;
    std::ostringstream text2;
    text1 << "| " << std::right << std::setw(width_index) << line.index << " |";
    text2 << "| " << std::left << std::setw(width_path) << unit->path() << " ";
    text2 << "| " << std::left << std::setw(width_type) << unit->type() << " ";
    text2 << "| " << std::left << std::setw(width_link) << line.text2 << " |";
    line.text1 = text1.str() + " ";
    line.text2 = " " + text2.str();
  }

  // Create table header.
  {
    std::ostringstream header;
    header << "| " << std::left << std::setw(width_index) << title_index << " ";
    header << "| " << std::left << std::setw(width_level) << title_level << " ";
    header << "| " << std::left << std::setw(width_path) << title_path << " ";
    header << "| " << std::left << std::setw(width_type) << title_type << " ";
    header << "| " << std::left << std::setw(width_link) << title_link << " ";
    header << "|";
    header_ = header.str();

    border_ += "|" + std::string(width_index + 2, '-');
    border_ += "|" + std::string(width_level + 2, '-');
    border_ += "|" + std::string(width_path + 2, '-');
    border_ += "|" + std::string(width_type + 2, '-');
    border_ += "|" + std::string(width_link + 2, '-');
    border_ += "|";
  }
}

void DumpNode::on_update(DiagGraph::ConstSharedPtr graph)
{
  const auto text_level = [](DiagUnit::DiagnosticLevel level) {
    if (level == DiagUnit::DiagnosticStatus::OK) return "OK   ";
    if (level == DiagUnit::DiagnosticStatus::WARN) return "WARN ";
    if (level == DiagUnit::DiagnosticStatus::ERROR) return "ERROR";
    if (level == DiagUnit::DiagnosticStatus::STALE) return "STALE";
    return "-----";
  };

  std::cout << border_ << std::endl;
  std::cout << header_ << std::endl;
  std::cout << border_ << std::endl;

  for (const auto & unit : graph->units()) {
    const auto & line = table_.at(unit);
    std::cout << line.text1 << text_level(unit->level()) << line.text2 << std::endl;
  }
}

}  // namespace diagnostic_graph_utils

#include <rclcpp_components/register_node_macro.hpp>
RCLCPP_COMPONENTS_REGISTER_NODE(diagnostic_graph_utils::DumpNode)
